src/magic_folder/magic_folder.py

```diff
@@ -918,3 +962,92 @@ def add_file(self, path):
         d = defer.Deferred()
         self._queue.put((bytespath, d))
         return d
+
+@attr.s
+@implementer(service.IService)
+class UploaderService(service.Service):
+    """
+    A service that periodically polls the database for local snapshots
+    and commit them into the grid.
+    """
+    _snapshot_store = attr.ib()
+    _local_author = attr.ib()
+    _tahoe_client = attr.ib()
+    _clock = attr.ib()
+    _polling_interval = attr.ib()
+
+    def startService(self):
+
+        service.Service.startService(self)
+
+        # do a looping call that polls the db for LocalSnapshots.
+        self._processing_loop = task.LoopingCall(
+            self._upload_local_snapshots,
+            self._tahoe_client,
+        )
+        self._processing_loop.clock = self._clock
+        self._processing = self._processing_loop.start(self._polling_interval, now=True)
+
+    def stopService(self):
+        """
+        Stop the uploader service.
+        """
+        service.Service.stopService(self)
+        d = self._processing
+        self._processing_loop.stop()
+        self._processing = None
+        self._processing_loop = None
+        return d
+
+
+    @eliotutil.inline_callbacks
+    def _upload_local_snapshots(self, tahoe_client):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+
+        :param tahoe_client: a TahoeClient instance
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._snapshot_store.get_all_item_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
```

Comment:
Yea, I think this is right.  Instead of loading a whole tree of snapshots from each row in the db, it should load a single snapshot from a single row.  The schema will probably change a bit (path no longer primary key) but those details can be hidden inside the database layer.  This code can ask for an iterator of snapshots or it can ask for the newest snapshot with no other local parents or something similar, send one off for processing, repeat.